students/allen_maxwell/Lesson_08/circle.py

```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env python3
+
+# Allen Maxwell
+# Python 210
+# 1/3/2020
+# circle.py
+
+import math
+
+
+class Circle(object):
+
+    def __init__(self, radius=0):
+        if not isinstance(radius, (int, float)):
+            raise ValueError('Radius must be a number')
+        elif radius < 0:
+            raise ValueError('Radius must be a positive number')
+        else:
+            self._radius = radius
+
+    @property
+    def radius(self):
+        return self._radius
+
+    @radius.setter
+    def radius(self, radius):
+        if not isinstance(radius, (int, float)):
+            raise ValueError('Radius must be a number')
+        elif radius < 0:
+            raise ValueError('Radius must be a positive number')
+        else:
+            self._radius = radius
+
+    @property
+    def diameter(self):
+        return self._radius * 2
+
+    @diameter.setter
+    def diameter(self, diameter):
+        if not isinstance(diameter, (int, float)):
+            raise ValueError('Diameter must be a number')
+        elif diameter < 0:
+            raise ValueError('Diameter must be a positive number')
+        else:
+            self._radius = diameter / 2
+
+    @property
+    def area(self):
+        return math.pi * self._radius * 2
+
+    @area.setter
+    def area(self, area):
+        raise AttributeError('Area cannot be set')
+
+    @classmethod
+    def from_diameter(cls, diameter):
+        if not isinstance(diameter, (int, float)):
+            raise ValueError('Diameter must be a number')
+        elif diameter < 0:
+            raise ValueError('Diameter must be a positive number')
+        else:
+            return cls(diameter / 2)
+
+    def __str__(self):
+        return 'Circle with radius: {}'.format(self._radius)
+
+    def __repr__(self):
+        return 'Circle({})'.format(self._radius)
+
+    def __add__(self, other):
+        if isinstance(other, str):
+            raise ValueError('Value must be a number')
+        elif isinstance(other, (int, float)):
+            if self.radius + other < 0:
+                raise ValueError('Value must less than or equal to the radius')
+            else:
+                return self.__class__(self._radius + other)
+        else:
+            return self.__class__(self._radius + other.radius)
+
+    def __radd__(self, other):
+        if isinstance(other, str):
+            raise ValueError('Value must be a number')
+        elif isinstance(other, (int, float)):
+            if self.radius + other < 0:
+                raise ValueError('Value must less than or equal to the radius')
+            else:
+                return self.__class__(self._radius + other)
+        else:
+            return self.__class__(self._radius + other.radius)
```

Comment:
there is a lot of the same code in various functions - could you clean this up by using a helper function?

Reply:
I updated circle to include helper functions and added more tests to more fully test the code. https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/pull/652/commits/bedc495df0035be600e7a8f8b493141025bf5065